README.md
# ASCIILOSCOPE
An oscilloscope in ASCII

The objective of this utility is to create a real time graphical monitoring of changing data from a terminal. The target applications may be monitoring electrical signals.  Clearly this is not as representative as a true graphical display, but it allows a simple illustration of input changes.  For my purposes it will allow me to select multipliers and offsets adjustments for signal captures by sensors for other applications.  It is also a test bed for the development of CursesSprites and eventual PerlayStation Terminal Games Console.  The development of these have been temporarily paused while I improve my understanding of terminal applications development (I am thinking of eventually disposing of Curses altogether, so Curses is not used in this application).  See [CHANGES](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/CHANGES.md) for updates. Conversion to a GUI, vector tracing appplication is projected, and may only require minor changes.

Ihe display can be positioned and sized as needed. Multiple traces can be displayed. The keyboard is monitored for functions to be added as development continues.

![Vesion 0.09](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/asciiloscope%20dual%20trace.gif)

### Features and function
This can be illustrated using the screenshot below.
![Screenshot2](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.09.jpg)

* Tab = Makes next trace active if there are multiple traces
* q   = Quits the application
* p   = Pauses display and data capture for all traces
* s   = Stop/resume individual traces
* h   = hide/show traces
* a   = Auto levels, adjusts multipliers and offset so data fits in the screen
* &#8594;   = Speed up increase number of updates per second by 10
* &#8592;   = Slow down, reduce number of updates per second by 10
* &#8593;   = Shift up, increase offset so trace moves up the display
* &#8595;   = Shift down, decrease offset so trace moves down the display
* \*    = Inc Window, increase samples per full width of the screen
* \/    = Dec Window, decrease samples per full width of the screen
* \+    = Magnify, increase Y multiplier by 10%, zooming into trace
* \-    = Reduce, reduce multiplier by 10%, zooming out

### Adding traces

Traces can be loaded from an external file by `my $trace= new Trace(traceFile=>"arduino.trc") `. Files are structured like the contents of a hash.

```
# Example contents of a .trc file
description   =>"Sine Trace",
name          =>"sin",
dataWindow    =>50,
internals     =>{x=>1},
symbol        => "*",
colour        => "red bold",
source        => sub{
                 my $self=shift;
		         shift @{$self->{data}} if @{$self->{data}}>$self->{dataWindow};
		         $self->{internals}{x}=0 if $self->{internals}{x}>200;
		         push @{$self->{data}},sin (3.14*$self->{internals}{x}++/20);
		      },
```
* `desription` is a user friendly description of the function
* `name` is the name of the trace and needsto be unique
* `dataWindow` is the number of data points displayed in each frame
* `internals` are specific to the plot where the user may store the internal variables for the trace functions
* `symbol` is the symbol used for the plot
* `source->()` is the function that retrieves the next data point. For illustration examples of sin and cos traces are supplied. Typically the function would remove the oldest datapoint (using `shift`) and insert (using `push`) the newest one at the other end. May be more reasonable to put in example triangle, sawtooth, square wave. A future trace will be the internal "trigger". The main purpose of this function is to capture external data for plotting.
* `colour` is a string of formatting options separated by spaces, e.g. "bold red strikethrough".

Alternatively the parameters can be passed to `new Trace` e.g. the following example reads the trace from a file containing data. This trace is also icluded in the src folder in file.trc format.. 
```
# Example trace that reads data from a file
$traces{file}=new Trace(
  description   =>"File Reader",
  name          =>"file",
  dataWindow    =>50,
  internals     =>{fileName=>"test.data",},
  symbol        => "*",
  colour        => "magenta bold",
  source        => sub{
     my $self=shift;
     open ($self->{internals}{fh},'<', $self->{internals}{fileName}) unless exists $self->{internals}{fh};
     shift @{$self->{data}} if @{$self->{data}}>$self->{dataWindow};
     my $in=readline($self->{internals}{fh});
     push @{$self->{data}},$in;
     close $self->{internals}{fh} and delete $self->{internals}{fh} if eof($self->{internals}{fh})
   },
);
```

### Options
The display from version 0.09 onwards, the display setu up by creating a Display object.
```
my $display=new Display(       # display parameters
   showLogo   =>1,             # show scrolling logo or not
   showMenu   =>1,             # show menu or not
   enableColours=>1,           # enable colours
   refreshRate =>100,
);
```
### Creating a Trace Widget

Widgets are given ids, and updated depdendent on `$display->{refreshRate}`.  Multiple traces can be displayed in a widget.

```
# create a chart widget containing traces, and then "run" the scope
$display->chart({id=>"scope",row=>4,column=>8,height=>17,width=>50,
	            borderStyle=>"double",borderColour=>"bold blue",
	            title=>"ASCIIloscope Demo",titleColour=>"black on_yellow",
	            traces=>[@traces]});
$display->run("scope");
```

It is also possible to create multiple chart widgets. `$display->run(<widgetId list>)`  starts a loop that redraws the contents of the widgets listed by their ids.

A trace can also be added after a chart widget has been created. using $display->addTrace(<trace>,<widgetId>)
	
```
# add trace to existing 
$display->addTrace( new Trace(traceFile=>"ramp.trc") ,"scope");   
```

## Charting from sensors

The goal of this project had been to produce low dependency visual representation of data acquied from sensors. The arduino.trc  file example allows data to be read from an Arduino to be read into the Asciiloscope.
```
 description   =>"Arduino Reader",
 name          =>"arduino",
 dataWindow    =>50,
 internals     =>{ init=>my $dummy=eval {use Device::SerialPort::Arduino},
                 },
 symbol        => "*",
 colour        => "yellow bold",
 source        => sub{
                  my $self=shift;
                  unless(defined $self->{internals}{device}){
                     $self->{internals}{device}=Device::SerialPort::Arduino->new(
                     port     => '/dev/ttyUSB0',
                     baudrate => 9600,
                     databits => 8,
                     parity   => 'none',
                     )
                   };
                  shift @{$self->{data}} if @{$self->{data}}>$self->{dataWindow};
                  my $in=$self->{internals}{device}->receive();
                  if ($in !~/\d/){$in=150+400*rand()};
                  push @{$self->{data}},$in-300;
           },
```


### Customisation

It is possible to customise the keyboard actions, either to replace existing actions, create new key actions, or to add functionailty to current action. Each key press is detected and the number corresponding to the key press is used to access  the action (stored in %actions hash in the package Display). `addAction` takes a key number and a hashref.  Hashref Keys are `note` for the note that appears on the menu, `proc` for the subroutine that is run when the key press is detected, and an optional `userProc` for user-defined actions. e.g.

```
# add/ modify action to a key press..tabs previously just changed activeTrace
# now it also updates the statBox
$display->addAction(9,{userProc=>sub{
     $display->statBox("stat");
  }
});   
```

###  Dependencies
* Time::HiRes
* Term::ReadKey;  
* ANSI compatible terminal

### Things to do: -
See [CHANGES](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/CHANGES.md) for updates.

* Manual scale and offset adjustment (done in v0.02)
* Manual sample rate adjustment      (done in v0.02)
* Autoscaling is already implemented (in v0.02 can now be triggered any time)
* Drift adjustment                   (done in v0.04)
* Colour                             (done in v0.05 and improved in 0.07)
* Multiple independent traces        (done in v0.05)
* Triggers and Storage modes         
* Freeze frame                       (done in v0.07) 
* Export data to CSV and graphically to SVG
* Data anaylsis
* OO redesign                        (done in 0.09)




CHANGES.md
### Version 0.10
- A Data analysis widget has been implementated
- Traces can be now loaded into an existing widget 
- The key-press actions can be modified from within main:: by using $display->addAction(<$OrdKey>,<subroutine ref>))
  
![Vesion 0.10](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.10.jpg)

### Version 0.09
  
- A virtually complete rewrite into a more object orientated design. Packages are however 
contained within the script as opposed to being loaded from external files, for convenience.
- Logo at bottom now replaced by a Logo scrolling in from the side at the beginning
- Traces are now loadable from external files
- Trace Windows can have titles appended
- Multiple Asciiloscope screens are possible

![Vesion 0.09](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.09.jpg)

![Vesion 0.09](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/asciiloscope%20dual%20trace.gif)

### Version 0.07
- Traces are individually coloured
- Traces may be hidden individually
- Pause/resume is toggled using 'P' (rather than using 'p' for pause and 'r' for resume)
- Stop/go for individual traces
- Trace names in bottom bar coloured as well
- Bug that caused glitches if colour not specified for trace fixed

![Vesion 0.07](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.07.jpg)

### Version 0.06
- Multiple traces now possible
- Possible to individually scale and shift traces.  Tab to toggle active Trace
- Trace names in bottom bar

![Vesion 0.06](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.06.jpg)

### Version 0.05
- Start colorisation, prepare for multiple traces
- Allow enabling/disabling of colour, menu and logo

![Vesion 0.05](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.05.jpg)

### Version 0.04
- Sample Window (@list size) can be increased or decreased
- Fix to OrdKey processing as advised by cybernicus

![Version 0.04](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.04.jpg)

### Version 0.03

- Freeze and Resume options added.
- Automatic clipping
- Zero line added
- quit does not leave cursor on plotarea

![Version 0.03](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/Version%200.03.jpg)


### Version 0.01

![Version 0.01](https://github.com/saiftynet/ASCIILOSCOPE/blob/master/images/asciiloscope.gif)


src/asciiloscope.pl
#!/usr/env/perl
#                      ASCIILOSCOPE
# A Terminal based real-time analog data visualisation tool.
# for more information visit:- https://github.com/saiftynet/ASCIILOSCOPE
# This program ANSI terminal escape codes for positioning and colours
# Not all terminals can handle these, sadly.

use strict;use warnings;

initialise();   # This function introduces the Trace and Display Package
                # as a subroutine.  one could put thes packages into
                # pm files, but included in this form for convenience

my $display=new Display(       # display parameters
   showLogo   =>0,             # show scrolling logo or not
   showMenu   =>1,             # show menu or not
   enableColours=>1,           # enable colours
   refreshRate =>100,
);

# create a chart widget containing traces
$display->chart({id=>"scope",row=>2,column=>8,height=>15,width=>50,
              borderStyle=>"double",borderColour=>"bold blue",
              title=>"ASCIIloscope Demo",titleColour=>"black on_yellow",
              traces=>[new Trace(traceFile=>"sin.trc") ] }); 

# create a statistics widget
$display->statBox({id=>"stat",row=>20,column=>8,height=>2,width=>50,
              borderStyle=>"double",borderColour=>"bold red",
              title=>"Statistics",titleColour=>"black on_yellow",});  

# add/ modify action to a key press..tabs previously just changed activeTrace
# now it also updates the statBox
$display->addAction(9,{userProc=>sub{
     $display->statBox("stat");
  }
});   

# add trace to existing 
$display->addTrace( new Trace(traceFile=>"ramp.trc") ,"scope");   

$display->run("scope");

exit;


## the following initialises twa packages Traces for handling the traces and
## Display that handles the display.  Ordinarily these would be loaded from .pm
## filed at a convenient @INC location, using `use`, but contained in this file
## for the time being for convenience.

sub initialise{
  
package Trace;
  use strict; use warnings;
  
  our $VERSION    = '0.10';
  
  sub new{
    my ($class, %args) = @_;
    my $self={};
    $self->{trace}={};
    foreach (keys %args){
         $self->{$_}=$args{$_} unless $_ eq "traceFile";
      }
    bless ($self,$class);
    loadTrace($self, "./".$args{traceFile}) if  $args{traceFile}; #and -e $args{traceFile};
    return $self;  
  }
  
  sub help{
    print "help\n";
  }
  
  sub loadTrace{
  my ($self,$traceFile)=@_;
  die "file $traceFile not found $!" unless -e $traceFile;
  my %import=do "./$traceFile" or die "Failed to load external traces $traceFile $!";
    foreach (keys %import){
      $self->{$_}=$import{$_};
    }
    $self->{data}=[] unless $self->{data};
    $self->nextData while @{$self->{data}}<$self->{dataWindow};
  }  
  
  sub nextData{
  my $self=shift;
  $self->{source}($self);  
  }
  sub setPlotArea{
   my ($self,$height,$width)=@_;
   $self->{height}= $height;
   $self->{width} = $width ;
     $self->autoLevels();
  }
  
  sub scatterPlot{
  my ($self,$plotArea)=@_;
  $plotArea//=[(" " x $self->{width}) x $self->{height}];
  return $plotArea if $self->{hidden};
  my @plots=map {$self->{data}[$_]?
      [int( $_*$self->{xMult}) ,
      bound (int($self->{yMult}*$self->{data}[$_] +$self->{yOffset}-.5),
         0,$self->{height}-1)]:()
       } (0..(@{$self->{data}}-1));
  foreach (@plots){
      substr ($plotArea->[$$_[1]], $$_[0],1,$self->{symbol}) if $$_[1]<$self->{height} and $$_[0]<$self->{width};
    }     
  return $plotArea
  }
  
  sub zeroLine{
  my ($self,$plotArea)=@_;
  my $zeroLine=bound($self->{yOffset}-0.5,0,$self->{height}-1);
  $plotArea->[$zeroLine]=~s/ /-/g;
  return $plotArea
  }
  
  sub colourise{
  my ($self,$plotArea)=@_;
  my $colour=$display->colour($self->{colour});my $reset=$display->colour("reset");
  foreach my $row (0..@$plotArea-1){
       $$plotArea[$row]=~s/([$self->{symbol}]+)/$colour$1$reset/g;
     }
  return $plotArea;
  }
  
  sub autoLevels{
    my $self=shift;
    $self->stats();
    $self->{yMult}=($self->{height}-2)/($self->{stats}{max}-$self->{stats}{min});
    $self->{yOffset}=-$self->{stats}{min}*$self->{yMult}+1;
    $self->{xMult}=$self->{width}/(scalar @{$self->{data}});
  };
 
 sub stats{
  my $self=shift;
  unless ($self->{data}->[0]){
    $self->nextData while @{$self->{data}}<$self->{dataWindow};
  }
    $self->{stats}{max}=$self->{data}->[0];$self->{stats}{min}=$self->{data}->[0]; $self->{stats}{sum}=0;
    my $squares=0;
    foreach my $y (@{$self->{data}}){
    next unless $y;
      $self->{stats}{max}=$y if  $y>$self->{stats}{max};
      $self->{stats}{min}=$y if  $y<$self->{stats}{min};
      $self->{stats}{sum}+=$y;
      $squares+=$y**2;
    } 
    $self->{stats}{mean}=$self->{stats}{sum}/$self->{dataWindow};
    $self->{stats}{rms}=sqrt ($squares/$self->{dataWindow});
    
 }
 
 sub bound{  
  my ($number,$min,$max)=@_;
  return $max if $number>$max;
  return $min if $number<$min;
  return $number;  
  }  
  
  sub getset{
    my ($self,$key,$value)=@_;
    $self->{$key}=$value if $value;
    return $self->{$key};
  }    
    
1;
package Display;
use strict; use warnings;
use Time::HiRes ("sleep");      # allow fractional sleeps 
use utf8;                       # allow utf characters in print
binmode STDOUT, ":utf8";
use Term::ReadKey;              # allow reading from keyboard
our $key;        
our @traceNames;
our $activeTrace;
our $lastTrace;

our $VERSION    = '0.10';

our %actions=(                  # for keyboard driven actions
   9=>{ # tab makes next trace active
     note=>"Tab = next trace",
     proc=>sub{my ($self,$tr)=@_;
       $lastTrace=$activeTrace;
       $activeTrace=pop @traceNames;
       unshift @traceNames,$activeTrace }
   },
   104=>{ # h hides trace
     note=>"h = Hide/show",
     proc=>sub{my ($self,$tr)=@_;$$tr{hidden}=!$$tr{hidden};},
   },
   113=>{ # q exits
     note=>"q = Exit",
     proc=>sub{ printAt(20,0,"Goodbye!");exit;},
   },
   112=>{  # Pause display (loop continues so keyboard is read)
     note=>"p = Pause/Resume",
     proc=>sub{my ($self,$tr)=@_;$self->{pause}=!$self->{pause}},
   },
   115=>{  # stop individual trace
     note=>"s = Stop trace/go",
     proc=>sub{my ($self,$tr)=@_;$$tr{frozen}=!$$tr{frozen}},
   },
   97 =>{  # Auto levels based on the current contents of @list
     note=>"a = Auto levels",
     proc=>sub{my ($self,$tr)=@_;$self->{traces}{$activeTrace}->autoLevels()},
   },
   67 =>{  # increase sample rate by 10
     note=>"🠞 = Speed up",
     proc=>sub{my ($self,$tr)=@_;$self->{refreshRate}+=10;},
   },
   68 =>{  # reduce sample rate by 10
     note=>"🠜 = Slow down",
     proc=>sub{my ($self,$tr)=@_;$self->{refreshRate}=$self->{refreshRate}>10?$self->{refreshRate}-10:10;},
   },
   65 =>{  # shift display up by 1
     note=>"🠉 = Shift up",
     proc=>sub{my ($self,$tr)=@_;$self->{traces}{$activeTrace}{yOffset}+=1;},
   },
   66 =>{ # shift display down by 1
     note=>"🠋 = Shift down",
     proc=>sub{my ($self,$tr)=@_;$self->{traces}{$activeTrace}{yOffset}-=1;},
   },   
   42 =>{ # increase samples per full width0..(@{$currentTrace{data}}
     note=>"* = Inc Window",
     proc=>sub{my ($self,$tr)=@_;unshift @{$self->{traces}{$activeTrace}{data}},($self->{traces}{$activeTrace}{data}[0]) x 10;
             $self->{traces}{$activeTrace}{xMult}=$$tr{width}/(scalar @{$self->{traces}{$activeTrace}{data}});},
   },
   47 =>{ # decrease samples per full width
     note=>"/ = Dec Window",
     proc=>sub{my ($self,$tr)=@_;@{$$tr{data}}=@{$$tr{data}}[9..@{$$tr{data}}] if  @{$$tr{data}}>15;
             $$tr{xMult}=$$tr{width}/( @{$$tr{data}});},
   },
   43 =>{ # increase multiplier by 10%
     note=>"+ = Magnify",
     proc=>sub{my ($self,$tr)=@_;return unless $activeTrace;$self->{traces}{$activeTrace}{yMult}*=1.1;},
   },
   45 =>{ # reduce multiplier by 10%
     note=>"- = Reduce",
     proc=>sub{my ($self,$tr)=@_;$self->{traces}{$activeTrace}{yMult}*=0.9;},
   },
);

our %borders=(
  simple=>{tl=>"+", t=>"-", tr=>"+", l=>"|", r=>"|", bl=>"+", b=>"-", br=>"+",},
  double=>{tl=>"╔", t=>"═", tr=>"╗", l=>"║", r=>"║", bl=>"╚", b=>"═", br=>"╝",},
  thin  =>{tl=>"┌", t=>"─", tr=>"┐", l=>"│", r=>"│", bl=>"└", b=>"─", br=>"┘",},  
  thick =>{tl=>"┏", t=>"━", tr=>"┓", l=>"┃", r=>"┃", bl=>"┗", b=>"━", br=>"┛",}, 
);

sub new{
    my ($class, %args) = @_;
    my $self={};
    $self->{traces}={};
    foreach (keys %args){
         $self->{$_}=$args{$_}
      }
      $self->{widgets}={};
      bless ($self,$class);
      $self->logo()  if $self->{showLogo};
      $self->menu()  if $self->{showMenu};
      return $self;
}

sub printAt{
  my ($row,$column,@textRows)=@_;
  @textRows = @{$textRows[0]} if ref $textRows[0];  
  my $blit="\033[?25l";
  $blit.= "\033[".$row++.";".$column."H".$_ foreach (@textRows) ;
  print $blit;
  print "\n"; # seems to flush the STDOUT buffer...if not then set $| to 1 
};

sub box{
  my ($self,$widgetData)=@_;
  $self->addWidget("box",$widgetData) unless $$widgetData{type};
  my %args=%$widgetData;
  $args{borderStyle}//="simple";
  my ($colour, $reset)=("","");
  if ($args{borderColour}){
    $colour=$self->colour($args{borderColour});
    $reset =$self->colour("reset");
  }
  my %border=%{$borders{$args{borderStyle}}};
  printAt ($args{row}-1,$args{column}-1,
      $colour.$border{tl}.($border{t} x $args{width}).$border{tr},
      ($border{l}.(" " x $args{width}).$border{r}) x $args{height},
      $border{bl}.($border{b} x $args{width}).$border{br}.$reset );
  if ($$widgetData{title}){
    printAt($args{row}-1,$args{column}+2,$self->colour($$widgetData{titleColour})." ".$$widgetData{title}." ".$self->colour("reset"));
  }
  print "\033[". 23 .";". 0 . "H";
}

sub chart{
  my ($self,$widgetData)=@_;
  $widgetData=$self->{widgets}{$widgetData} unless ref $widgetData;
  my %args=%$widgetData;
  unless ($$widgetData{type}){ # if widget not previously added
    $self->box($widgetData);
    my $legendRow=$args{height}+$args{row};
    my $legendColumn=$args{column}+2;
    foreach my $trace (@{$args{traces}}){
      $trace->setPlotArea($args{height},$args{width});
      printAt($legendRow,$legendColumn," ".$self->colour($trace->{colour}).$trace->{symbol}." ".$trace->{name}." ");
      $legendColumn+=5+length $trace->{name};
      $self->{traces}{$trace->{name}}=$trace;
      push @traceNames,$trace->{name};
      $activeTrace=$trace->{name};
      die unless defined $self->{traces}{$activeTrace}{yMult};
    } ;
    $$widgetData{started}=1;
    $self->addWidget("chart",$widgetData);
  };
  my $plot;
    unless ($self->{pause}){
    undef $plot;
    return unless @{$args{traces}};
    foreach my $trace (@{$args{traces}}){
      $plot= $trace->scatterPlot($plot) ;
      $plot= $trace->zeroLine ($plot) if $trace->{name} eq $activeTrace;
      $trace->nextData() unless $trace->{frozen};
    }
    foreach my $trace (@{$args{traces}}) {
      $plot=$trace->colourise($plot)
    }
    printAt ($args{row},$args{column}, reverse @$plot);
  }
}  

sub statBox{  #a widget that displays the statistics for current widget
  my ($self,$widgetData)=@_;
  $widgetData=$self->{widgets}{$widgetData} unless ref $widgetData;
  my %args=%$widgetData;
  unless ($$widgetData{type}){
     $self->box($widgetData);
    }
  printAt ($args{row},$args{column},       
     map{$_?statString($self,$_):()}($activeTrace,$lastTrace) );
     
     
  sub statString{
    my ($self,$traceName)=@_;    
    $self->{traces}{$traceName}->stats();
    return colour($self,$self->{traces}{$traceName}{colour}).($self->{traces}{$traceName}{name}).colour($self,"reset").
     " max=" .(sprintf "%.3g",$self->{traces}{$traceName}{stats}{max}).
     " min=" .(sprintf "%.3g",$self->{traces}{$traceName}{stats}{min}).
     " mean=".(sprintf "%.3g",$self->{traces}{$traceName}{stats}{mean}).
     " rms=" .(sprintf "%.3g",$self->{traces}{$traceName}{stats}{rms})
    
  }
}

sub addAction{
  my ($self,$key, $actionData)=@_;
  my %args=%$actionData;
  foreach my $k (keys %args){
    $actions{$key}{$k}=$args{$k};
  }
}

sub addTrace{
  my ($self, $trace, $widget)=@_;
  push @{$self->{widgets}{$widget}{traces}},$trace;
  foreach my $tr (@{$self->{widgets}{$widget}{traces}}){
    foreach  (0..$#traceNames){
      if ($tr->{name} eq $traceNames[$_]){
        splice @traceNames, $_, 1;
        last;
      }
    }
  }
  $self->{widgets}{$widget}{type}=undef;
  $self->chart($widget)
}

sub run{
  my ($self,@widgetIds)=@_;
  ReadMode 'cbreak';
  while(1){
    foreach my $id(@widgetIds){
    if (exists $self->{widgets}->{$id}){
      if ($self->{widgets}->{$id}->{type} eq "chart"){$self-> chart($id)} 
    }
    }
  sleep 1/$self->{refreshRate};
  $key = ReadKey(-1);                # -1 means non-blocking read
  if ($key){
    my $OrdKey = ord($key);
    printAt (20,60,"key pressed=$OrdKey");
    # Keys actions are stored in %actions
    $actions{$OrdKey}{proc}->($self,$self->{traces}{$activeTrace}) if defined $actions{$OrdKey}{proc} and $activeTrace;
    $actions{$OrdKey}{userProc}->($self,$self->{traces}{$activeTrace}) if defined $actions{$OrdKey}{userProc};    
  }
  }  
  ReadMode 'normal';  
}  

sub addWidget{
  my ($self,$type,$widgetData)=@_;
  $$widgetData{type}=$type;
  $self->{widgets}->{$$widgetData{id}}=$widgetData;
}

sub drawWidget{
  my ($self,$widgetData)=@_;
  if ($$widgetData{type}=="box") { $self->box($widgetData); };
}

sub scrollIn{
  my ($row,$column,$width,@message)=@_;
  my $at=$column+$width;my $start=0;
  my $l=length $message[0];
  foreach my $section(0..$width+$l){
    my @sliced=();
    $start++ if $section>$width;
    my $length=$section<$width?$section:($width+$l-$section+1)>$width?$width:($width+$l-$section+1);
    
    foreach my $mRow (@message){
    push @sliced,(substr $mRow." ",$start,$length);
     }
   printAt ($row,$at, @sliced);  
   $at-- unless $at==$column;
     sleep 0.02;
    }
}

sub logo{
  my $self=shift;
  print $self->colour("bold yellow");
  scrollIn(8,8,70,
    '   _    ___   __  _______  _    ___   ___   __   ___   ___  ___ ',
    '  /_\  / __| / _||_ _|_ _|| |  / _ \ / __| / _| / _ \ |  _\| __|',
    ' / _ \ \__ \| (_  | | | | | |_| (_) |\__ \| (_ | (_) || |_/| _| ',
    '/_/ \_\|___/ \__||___|___||___|\___/ |___/ \__| \___/ |_|  |___|');
    print $self->colour("reset");
}

sub menu{
  my $self=shift;
  # Print Menu...disable by setting $display{showMenu} to zero
    printAt( 2,62, $self->colour("underline")."MENU".$self->colour("reset"),map{$actions{$_}{note} } sort { $a <=> $b } keys %actions);
}

## colour handling routines....could be more elegantly done iusing Term::ANSIColor
## This is a more compact version

sub stripColours{
  my $line=shift;
  $line=~s/\033\[[^m]+m//g;
  return $line;
}

sub colour{
  my ($self,$fmts)=@_;
  return "" unless $fmts and $self->{enableColours}; 
  my @formats=map {lc $_} split / +/,$fmts;   
  my %colours=(black   =>30,red   =>31,green   =>32,yellow   =>33,blue   =>34,magenta   =>35,cyan  =>36,white   =>37,
               on_black=>40,on_red=>41,on_green=>42,on_yellow=>43,on_blue=>44,on_magenta=>4,on_cyan=>46,on_white=>47,
               reset=>0, bold=>1, italic=>3, underline=>4, strikethrough=>9,);
  return join "",map {defined $colours{$_}?"\033[$colours{$_}m":""} @formats;
}

1;
}
